### ISSUE TITLE
Libp2p daemon integration

### ISSUE THREAD

**Issue by mhchia** [ethereum/trinity#293]
### What is wrong?

Per the libp2p requirement in https://github.com/ethereum/trinity/issues/281, we will require the code to communicate with [libp2p daemon](https://github.com/libp2p/go-libp2p-daemon). I have some questions for future development:
- Should I make the PRs to another branch, e.g. libp2p-daemon?
- To interact with libp2p-daemon, I use [mhchia/py-libp2p-daemon-bindings](https://github.com/mhchia/py-libp2p-daemon-bindings) to do it. Should I move this package under the package `p2p`, or somewhere else, instead of import it as a third-party package?
- After roughly tracing the code, it seems I need to modify the package `p2p`(for the `Peer`, `PeerPool`, etc...), and also to add data structures for the shard chains. Does it make sense to add these data structures under `eth2`, e.g. a new directory `shards`?
- Does it make sense to add integration tests, with the go-libp2p-daemon spun up?

Thanks a lot!

**Comment by cburgdorf** [comment]
Cool! Here are my thoughts:

>* Should I make the PRs to another branch, e.g. libp2p-daemon?

PRs against `master` seem fine to me.

>* Should I move this package under the package `p2p`, or somewhere else, instead of import it as a third-party package?

Mmh. Depends, would it make our life much harder if we used it as a `pip` dependency? Does it have to be updated very frequently?

>* I need to modify the package `p2p`(for the `Peer`, `PeerPool`, etc...),

You'll probably need some modifications directly in `p2p` but I *think* I'd recommend we try to keep things mostly out of the `p2p` package what doesn't absolutely need to be there. Depending on the scope (datastructure needed in `eth2` vs just trinity machinery) things should either go directly into `eth2` or somewhere in trinity. Where the *somewhere* in Trinity is a bit up in the air if you ask me. If it's stuff that can sit somewhere symmetrical with some existing types (e.g. `trinity.protocol.*`) then it should go there, if it is entirely new stuff then maybe something like `trinity.libp2p` could serve as a temporary bucket for the lack of a better place?

> * Does it make sense to add integration tests, with the go-libp2p-daemon spun up?

Sounds useful!

I'm sure the rest of us have thoughts on this as well.

**Comment by mhchia** [comment]
> Mmh. Depends, would it make our life much harder if we used it as a `pip` dependency? Does it have to be updated very frequently?

The bindings code is subject to changes. In this case, it seems more suitable to move the code to trinity's codebase?

> I'd recommend we try to keep things mostly out of the `p2p` package what doesn't absolutely need to be there. Depending on the scope (datastructure needed in `eth2` vs just trinity machinery) things should either go directly into `eth2` or somewhere in trinity. If it's stuff that can sit somewhere symmetrical with some existing types (e.g. `trinity.protocol.*`) then it should go there, if it is entirely new stuff then maybe something like `trinity.libp2p` could serve as a temporary bucket for the lack of a better place?

Agree. It seems pretty reasonable.

Really appreciate your answers!:)

**Comment by pipermerriam** [comment]
I think keeping this fully isolated from the existing code may be best but that's just my inclination without having looked at any actual details.  That would imply a new top level module `libp2p`

**Comment by mhchia** [comment]
> That would imply a new top level module `libp2p`

Sounds reasonable to me! I can start coding based on this structure and change the structure if needed in the future. Thanks a lot for the answer:)

**Comment by ralexstokes** [comment]
yeah following up on what @pipermerriam is saying @mhchia there are (at least) two stages here.

1. integration of the libp2p client (these daemon bindings) which this issue addresses
2. use of libp2p for the beacon node p2p layer (separate task, unclear if there is an issue to track this... I'll go open one if there is not).

i would suggest we expose a `libp2p.Host` class at the top-level of this new `libp2p` module which for now will wrap the `p2pclient.Client`. once the Python-native libp2p has landed, we can just swap out the proxy `libp2p.Host` for the real `libp2p.Host` (avoiding the daemon)

we now have the separate task of designing how the `BeaconNode` will talk to this `libp2p` service which i'll open an issue to discuss. and I'll go leave a comment on your PR to sketch what i mean above w/ the `libp2p.Host` shim

**Comment by mhchia** [comment]
> 1. integration of the libp2p client (these daemon bindings) which this issue addresses
> 2. use of libp2p for the beacon node p2p layer (separate task, unclear if there is an issue to track this... I'll go open one if there is not).

Sounds good to me. I think there is no issue tracking for the task (2). I'd like to list more detailed steps after the issue is opened(or I can open it if needed).

> i would suggest we expose a `libp2p.Host` class at the top-level of this new `libp2p` module which for now will wrap the `p2pclient.Client`. once the Python-native libp2p has landed, we can just swap out the proxy `libp2p.Host` for the real `libp2p.Host` (avoiding the daemon)

That will be great! I'm wondering what `libp2p.Host` should include. If mapped to libp2p, what `p2pclient.Client` with the libp2p daemon supports is roughly `Host`(including `Swarm`(network components), `peerstore`(data structure storing peer identifiers, ...), `DHT`, `PubSub`, `ConnectionManager`. If `libp2p.Host` wraps `p2pclient`, it means it includes all these functionalities. I think it is reasonable have this interface for this stage, and it should be easily assembled by the native py-libp2p.

> we now have the separate task of designing how the `BeaconNode` will talk to this `libp2p` service which i'll open an issue to discuss. and I'll go leave a comment on your PR to sketch what i mean above w/ the `libp2p.Host` shim

That will be great to have the issue to confirm next things to be done. Thanks a lot!

> after thinking a bit more, i think it makes more sense to push this `libp2p.Host` into some beacon node networking code, leaving the `libp2p` module as is on this front -- i expect we will just add `py-libp2p` as a dependency once it is ready and if we craft this interface properly it should be easy to swap out

I'm fine with this structure:)

**Comment by ralexstokes** [comment]
@mhchia yeah -- i'm looking into some stuff around using [`curio`](https://github.com/dabeaz/curio) after @pipermerriam mentioned it could be worthwhile -- from what i have seen it looks like it gives us really nice concurrency semantics and i have a vague desire to take a stab at what the beacon chain infra could look like if we build it around this framework. i feel like i should have more of an idea together before i open an issue for discussion however. and don't let me stop you if you wanna move ahead :)

**Comment by ralexstokes** [comment]
@mhchia ok to close this? we have moved ahead w/ our fancy 'native' py-libp2p implementation

**Comment by mhchia** [comment]
Yeah sure😂
